### PR TITLE
Fix leaky bucket leak calculation

### PR DESCRIPTION
**Pull request by seancaffery**
Fixes issue where the bucket does not leak when the time since last
update is less than the interval. This meant the bucket never leaked
and was only emptied if the limit wasn't checked for the entire interval

@zendesk/echidna @grosser

### CODE REVIEW

File: lib/prop/leaky_bucket_strategy.rb [hunk 1]
```diff
@@ -8,9 +8,10 @@ class << self
       def counter(cache_key, options)
         bucket = Prop::Limiter.cache.read(cache_key) || zero_counter
         now = Time.now.to_i
-        leak_amount = (now - bucket.fetch(:last_updated)) / options.fetch(:interval) * options.fetch(:threshold)
+        leak_rate = (now - bucket.fetch(:last_updated)) / options.fetch(:interval).to_f
```
Comment (seancaffery) [inline lib/prop/leaky_bucket_strategy.rb]: The problem is integer division here. `now - last_updated` is usually smaller than `interval` which results in `0 * threshold` -> 0

File: lib/prop/leaky_bucket_strategy.rb [hunk 2]
```diff
@@ -8,9 +8,10 @@ class << self
...
+        leak_amount =  leak_rate * options.fetch(:threshold)
 
-        bucket[:bucket] = [bucket.fetch(:bucket) - leak_amount, 0].max
+        bucket[:bucket] = [(bucket.fetch(:bucket) - leak_amount).to_i, 0].max
```
Comment (grosser) [inline lib/prop/leaky_bucket_strategy.rb]: now it can still be 0 ? <-> should we get rid of 0 and make it `, 1].max` ?

File: test/test_leaky_bucket_strategy.rb [hunk 3]
```diff
@@ -39,6 +39,22 @@
...
+    end
+
+    it "rounds down when applying leak amount to bucket" do
+      # this leaves the bucket at 5 rather than 6, leaning towards allowing more requests
+
+      # bucket updated 5 seconds ago with leak rate of 0.983/second
+      Prop::Limiter.cache.write(@key, { bucket: 10, last_updated: @time.to_i - 5 })
+      Prop::LeakyBucketStrategy.increment(@key, 0, interval: 60, threshold: 59)
+      Prop::Limiter.cache.read(@key).must_equal bucket: 5, last_updated: @time.to_i
+    end
```
Comment (grosser) [inline test/test_leaky_bucket_strategy.rb]: add a test here for the edge-case of decrementing by 0

File: lib/prop/leaky_bucket_strategy.rb (hunk 2 above)
Comment (seancaffery) [reply lib/prop/leaky_bucket_strategy.rb]: I think leaving 0 here is fine. `max` with 0 resets the bucket if the throttle hasn't been checked within the interval.

File: test/test_leaky_bucket_strategy.rb (hunk 3 above)
Comment (seancaffery) [reply test/test_leaky_bucket_strategy.rb]: done

File: test/test_leaky_bucket_strategy.rb [hunk 4]
```diff
@@ -52,6 +68,13 @@
       Prop::LeakyBucketStrategy.decrement(@key, 3, interval: 1, threshold: 10)
       Prop::Limiter.cache.read(@key).must_equal bucket: 2, last_updated: @time.to_i
     end
+
+    it "adjusts bucket when elapsed time is less than interval" do
+      # bucket updated 5 seconds ago with leak rate of 1/second
+      Prop::Limiter.cache.write(@key, { bucket: 10, last_updated: @time.to_i - 5 })
+      Prop::LeakyBucketStrategy.decrement(@key, 0, interval: 60, threshold: 60)
+      Prop::Limiter.cache.read(@key).must_equal bucket: 5, last_updated: @time.to_i
+    end
```
Comment (grosser) [inline test/test_leaky_bucket_strategy.rb]: I was hoping for a test that shows that the `bucket` value did not go down because we hit the `0` case ... or does that never happen ?

Comment (seancaffery) [reply test/test_leaky_bucket_strategy.rb]: oh gotcha. That can happen if the bucket is was last updated in the current second. Added a test.
